e2e/libs/page-manager.ts

```diff
@@ -0,0 +1,241 @@
+import { Browser, ConsoleMessage, launch, LaunchOptions, Page, Request } from 'puppeteer';
+import {
+  showFailedResponse,
+  describeJsHandle,
+  getRequestData,
+  isWorkbenchRequest,
+  logError,
+  hideResponse,
+  transformResponseBody
+} from './page-events-helper';
+import { logger } from './logger';
+import { defaultLaunchOptions } from './page-options';
+import fs from 'fs-extra';
+import { signIn, signInWithAccessToken } from 'utils/test-utils';
+import { savePageToFile, takeScreenshot } from 'utils/save-file-utils';
+
+const failScreenshotDir = 'logs/screenshot';
+const failHtmlDir = 'logs/html';
+const { CIRCLE_BUILD_NUM } = process.env;
+const userAgent =
+  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) ' +
+  'Chrome/80.0.3987.149 Safari/537.36' +
+  (CIRCLE_BUILD_NUM ? ` (circle-build-number/${CIRCLE_BUILD_NUM})` : '');
+
+/**
+ * Launch new Chrome.
+ * @param launchOpts: {@link LaunchOptions} New browser launch options.
+ * @return {@link Browser}
+ */
+export const launchBrowser = async (launchOpts?: LaunchOptions): Promise<Browser> => {
+  return launch(launchOpts ? launchOpts : defaultLaunchOptions);
+};
+
+/**
+ * Launch new incognito page.
+ * @param browser: {@link Browser}
+ * @return Page
+ */
+export const launchPage = async (browser: Browser): Promise<Page> => {
+  const incognitoPage = await browser.createIncognitoBrowserContext().then((context) => context.newPage());
+  await initPageBeforeTest(incognitoPage);
+  return incognitoPage;
+};
+
+export const withBrowser = (launchOpts?: LaunchOptions) => async (
+  testFn: (browser: Browser) => Promise<void>
+): Promise<void> => {
+  const browser = await launchBrowser(launchOpts);
+  try {
+    await testFn(browser);
+  } finally {
+    await browser
+      .close()
+      .then(() => logger.info('Closing browser'))
+      .catch((err: Error) => {
+        console.error(`Unable to close browser. Error message: ${err.message}`);
+      });
+  }
+};
+
+/**
+ * Launch new browser and incognito page. Opens Workbench Login page.
+ * @param launchOpts: {@link LaunchOptions} New browser launch options.
+ */
+export const withPage = (launchOpts?: LaunchOptions) => async (
+  testFn: (page: Page, browser: Browser) => Promise<void>
+): Promise<void> => {
+  await withBrowser(launchOpts)(async (browser) => {
+    const incognitoPage = await browser.createIncognitoBrowserContext().then((context) => context.newPage());
+    try {
+      await initPageBeforeTest(incognitoPage);
+      await testFn(incognitoPage, browser);
+    } catch (err) {
+      if (err instanceof Error) {
```

Comment:
Sorry.
Regarding to renaming `withPage`, what does `rwTest` stands for? I think `withPage` is a clear name and makes sense.

We're not saving pdf file right now because it only works in headless mode. It breaks test when debugging test in headful mode. It's there for future usage.

Since the `browser` and `page` objects are not global, there is no way to handle failures generically outside the `withPage` function.

Reply:
If there's no other place you can find to put this, then I guess this is fine.

What I am attempting to point out is that a function called `withPage()` I would expect to create and provide a page. It's doing this, but it's also handling a bunch of error catching / logging logic which all assume that the function is being used in a specific way - i.e. it would never make sense to use this function outside the context of executing a test. Ideally this is either handled somewhere else instead, or the function is named in a way that more accurately captures the semantics.

Some alternative ideas:

```
withPageTest()
doTestWithPage()
withHarness()
aouTest()
rwTest() // rw = Researcher Workbench
```

Reply:
Thanks. I renamed `withPage` to `withPageTest`.